File: Heat/Authorize/ResourceAuthorizeAttribute.vb
Imports Heat.Repositories
Imports Heat.Controllers
Imports Microsoft.Owin
Imports Microsoft.AspNet.Identity.Owin
Imports System.Security.Claims
Imports System.Web.Security
Imports System.Collections.ObjectModel




<AttributeUsage(AttributeTargets.Class Or AttributeTargets.Method, inherited:=True, AllowMultiple:=True)> _
Public Class ResourceAuthorizeAttribute
    Inherits AuthorizeAttribute

    Private _operation As ResourceOperations
    Private _resource As String

    Public Sub New(Operation As ResourceOperations, Resource As String)
        _operation = Operation
        _resource = Resource
    End Sub
     

    Public Overrides Sub OnAuthorization(filterContext As Mvc.AuthorizationContext)
        Dim dbContext = filterContext.HttpContext.GetOwinContext().Get(Of HeatDBContext)()
        Dim cm As New ResourceClaimsAuthorizeManager(dbContext)

        Dim p As ClaimsPrincipal = DirectCast(filterContext.HttpContext.User, ClaimsPrincipal)
        Dim resourceClaims As New Collection(Of Claim)
        Dim operationClaims As New Collection(Of Claim)

        resourceClaims.Add(New Claim("http://scheme.diegobrocchi.it/claims/2015/resource", _resource))
        operationClaims.Add(New Claim("http://scheme.diegobrocchi.it/claims/2015/operation", _operation, "ResourceOperations"))

        If cm.CheckAccess(New AuthorizationContext(DirectCast(filterContext.HttpContext.User, ClaimsPrincipal), resourceClaims, operationClaims)) Then
            MyBase.OnAuthorization(filterContext)
        Else
            MyBase.HandleUnauthorizedRequest(filterContext)

        End If

    End Sub

     

End Class

File: Heat.ConvertedToC#/Authorize/ResourceAuthorizeAttribute.cs
using System;
using System.Web;
using System.Web.Mvc;
using Heat.Repositories;
using Microsoft.AspNet.Identity.Owin;
using System.Security.Claims;
using System.Collections.ObjectModel;
namespace Heat
{




    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, Inherited = true, AllowMultiple = true)]
	public class ResourceAuthorizeAttribute : AuthorizeAttribute
	{

		private ResourceOperations _operation;

		private string _resource;
		public ResourceAuthorizeAttribute(ResourceOperations Operation, string Resource)
		{
			_operation = Operation;
			_resource = Resource;
		}


		public override void OnAuthorization(System.Web.Mvc.AuthorizationContext filterContext)
		{
			var dbContext = filterContext.HttpContext.GetOwinContext().Get<HeatDBContext>();
			ResourceClaimsAuthorizeManager cm = new ResourceClaimsAuthorizeManager(dbContext);

			ClaimsPrincipal p = (ClaimsPrincipal)filterContext.HttpContext.User;
			Collection<Claim> resourceClaims = new Collection<Claim>();
			Collection<Claim> operationClaims = new Collection<Claim>();

			resourceClaims.Add(new Claim("http://scheme.diegobrocchi.it/claims/2015/resource", _resource));
			operationClaims.Add(new Claim("http://scheme.diegobrocchi.it/claims/2015/operation", _operation.ToString() , "ResourceOperations"));

			if (cm.CheckAccess(new System.Security.Claims.AuthorizationContext((ClaimsPrincipal)filterContext.HttpContext.User, resourceClaims, operationClaims))) {
				base.OnAuthorization(filterContext);
			} else {
				base.HandleUnauthorizedRequest(filterContext);

			}

		}



	}
}
